Add `RSocketInboundGateway`; refactoring

artembilan opened the pull request:
* Extract an `AbstractRSocketConnector` for common client and server connectors logic
* Introduce an `IntegrationRSocketAcceptor` and `IntegrationRSocket` for the
mapping and handling logic of RSockets and messages in between
* Introduce an `IntegrationRSocketEndpoint` marker interface for Inbound Gateway mappings
* Add `RSocketInboundGateway` implementation, which is called from the
`IntegrationRSocketAcceptor` by the `IntegrationRSocketEndpoint` mapping
* Add `RSocketConnectedEvent` to emit when the client is connected to the server.
It does not make sense in Spring Integration to delegate such a logic into the
`RSocketInboundGateway`

<!--
Thanks for contributing to Spring Integration. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc).
-->

artembilan committed:
Add `RSocketInboundGateway`; refactoring

* Extract an `AbstractRSocketConnector` for common client and server connectors logic
* Introduce an `IntegrationRSocketAcceptor` and `IntegrationRSocket` for the
mapping and handling logic of RSockets and messages in between
* Introduce an `IntegrationRSocketEndpoint` marker interface for Inbound Gateway mappings
* Add `RSocketInboundGateway` implementation, which is called from the
`IntegrationRSocketAcceptor` by the `IntegrationRSocketEndpoint` mapping
* Add `RSocketConnectedEvent` to emit when the client is connected to the server.
It does not make sense in Spring Integration to delegate such a logic into the
`RSocketInboundGateway`

artembilan commented:
This is not fully ready yet.

At least tests are needed to cover an introduced functionality.

The server inbound side and docs might be deferred into a separate PR...

Right now I just want to share what I have so far to gather any possible feedback.

Thanks

garyrussell commented on spring-integration-rsocket/src/main/java/org/springframework/integration/rsocket/IntegrationRSocketAcceptor.java:
```diff
@@ -0,0 +1,132 @@
+	@Override
+	public Mono<RSocket> accept(ConnectionSetupPayload setupPayload, RSocket sendingRSocket) {
+		IntegrationRSocket rsocket = createRSocket(sendingRSocket);
+		rsocket.setApplicationEventPublisher(this.applicationEventPublisher);
```
null check or no-op default implementation.

garyrussell commented on spring-integration-rsocket/src/main/java/org/springframework/integration/rsocket/IntegrationRSocket.java:
```diff
@@ -0,0 +1,238 @@
+ *
+ * @see org.springframework.messaging.rsocket.MessagingRSocket
+ */
+class IntegrationRSocket extends AbstractRSocket {
```
Oh, I see it's not a bean.

garyrussell commented on spring-integration-rsocket/src/main/java/org/springframework/integration/rsocket/IntegrationRSocket.java:
```diff
@@ -0,0 +1,238 @@
+		int refCount = refCount(dataBuffer);
+		return Mono
+				.defer(() -> {
+					this.applicationEventPublisher.publishEvent(
```
See comment at caller.

garyrussell commented on spring-integration-rsocket/src/main/java/org/springframework/integration/rsocket/RSocketConnectedEvent.java:
```diff
@@ -0,0 +1,64 @@
+	public RSocketRequester getRequester() {
+		return this.requester;
+	}
+
```
I find `toString()` useful on events.

garyrussell reviewed (changes requested):
A few comments from first review pass.

artembilan committed:
* Add `ServerRSocketConnector` to represent an RSocket server and
container for connected `RSocketRequester`s from clients
* Extract `accept(ConnectionSetupPayload setupPayload, RSocket sendingRSocket)`
server logic into an internal `ServerRSocketAcceptor` extension for the
`IntegrationRSocketAcceptor`
* Address PR comments:
 - `RSocketConnectedEvent.toString()`
 - `ApplicationEventPublisherAware` into the `ServerRSocketConnector`
 - Log RSocket connection if no `this.applicationEventPublisher`
* Remove `handleConnectionSetupPayload()` from the `IntegrationRSocket`
since it is not delegated to the target handler
* Provide reasonable default `RSocketStrategies` for the
`AbstractRSocketConnector` and `RSocketInboundGateway`
* Add initial `RSocketInboundGatewayIntegrationTests`

artembilan commented:
Pushed some further implements and implementation.
See commit comment for details.

Thanks

garyrussell commented on spring-integration-rsocket/src/main/java/org/springframework/integration/rsocket/RSocketConnectedEvent.java:
```diff
@@ -61,4 +63,13 @@ public RSocketRequester getRequester() {
+	public String toString() {
+		return "RSocketConnectedEvent{" +
+				"destination='" + this.destination + '\'' +
+				", data=" + Arrays.toString(this.data.asByteBuffer().array()) +
```
I don't think rendering a `byte[]` as a bunch of ints is useful, maybe omit this (or just include the length, perhaps).

garyrussell reviewed (changes requested):
Just one issue from this pass.

artembilan committed:
* Don't use `BytyBuffer` in the `RSocketConnectedEvent.toString()`
* Always `payload.retain()` when we convert `Payload` to `DataBuffer`
* Fix `IntegrationRSocketAcceptor.detectEndpoints()` stream logic
to really iterate over all the `IntegrationRSocketEndpoint` beans
* Fix test to use an explicit `ClientConfig` class for the
`@SpringJUnitConfig`: looks like JUnit 5 is OK to scan for `package
protected` classes as well
* Add request/reply tests into the `RSocketInboundGatewayIntegrationTests`
for both server and client sides

artembilan commented:
OK. Now I think it is very close for merging.

Pay attention, please, how `@Transformer` looks in the `RSocketInboundGatewayIntegrationTests.CommonConfig`. It is nice to have it without subscription - everything is done by the RSocket protocol.

garyrussell commented on spring-integration-rsocket/src/main/java/org/springframework/integration/rsocket/RSocketConnectedEvent.java:
```diff
@@ -67,7 +65,7 @@ public RSocketRequester getRequester() {
 		return "RSocketConnectedEvent{" +
 				"destination='" + this.destination + '\'' +
-				", data=" + Arrays.toString(this.data.asByteBuffer().array()) +
+				", data length=" + this.data.asByteBuffer().array().length +
```
This still seems to be a lot of overhead in a `toString()`.

Why do we need this `DataBuffer` in the event?

artembilan committed:
* Remove `DataBuffer` from `RSocketConnectedEvent.toString()`

artembilan committed:
* Fix Checkstyle violation in the RSocketInboundGatewayIntegrationTests
